### PR TITLE
Add retry to AssertCleanup

### LINKED ISSUE #61484: [pr:pull-kubernetes-e2e-gce] Kubectl client [k8s.io] Update Demo should create and stop a replication controller  [Conformance] flaked 53 times in the past week
### Flaky Job: pr:pull-kubernetes-e2e-gce
[Consistency: 88.5%](http://storage.googleapis.com/k8s-metrics/flakes-daily-latest.json)

### Flakiest test : [sig-cli] Kubectl client [k8s.io] Update Demo should create and stop a replication controller  [Conformance]
[Flakes in the past week: 53](http://storage.googleapis.com/k8s-metrics/flakes-latest.json)

/priority failing-test
/priority important-soon
/sig cli
/kind flake

@kubernetes/sig-cli-test-failures

### PR DESCRIPTION
**Pull request by Liujingfang1**
**What this PR does / why we need it**:
Add retry in AssertCleanup for e2e kubectl test to remove the flakeness

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #61484

**Special notes for your reviewer**:
This function is only used in e2e kubectl test. It doesn't affect other tests' behavior.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by mengqiy** [comment]
@soltysh Can you approve this?

File: test/e2e/framework/util.go [hunk 1]
```diff
@@ -2066,16 +2066,29 @@ func AssertCleanup(ns string, selectors ...string) {
 	if ns != "" {
 		nsArg = fmt.Sprintf("--namespace=%s", ns)
 	}
+	var err error
+	for i := 1; i <= 3; i++ {
```
Comment (soltysh) [inline test/e2e/framework/util.go]: Wouldn't be better to use `wait.ExponentialBackoff` instead?

**Review by soltysh — APPROVED** [review]
The change itself lgtm, but I'd like to know if the suggested approach was considered.

File: test/e2e/framework/util.go (hunk 1 above)
Comment (Liujingfang1) [reply test/e2e/framework/util.go]: @soltysh I wasn't aware of the function `wait.ExponentialBackoff`. I took a look at it and it is a better approach. I will update the change. Thank you for your suggestion.

File: test/e2e/framework/util.go [hunk 2]
```diff
@@ -2066,15 +2066,28 @@ func AssertCleanup(ns string, selectors ...string) {
...
+	backoff := wait.Backoff{
+		Duration: 5 * time.Second,
+		Factor:   2,
+		Steps:    3,
+	}
+	verifyCleanupFunc := func() (bool, error) {
+		for _, selector := range selectors {
+			resources := RunKubectlOrDie("get", "rc,svc", "-l", selector, "--no-headers", nsArg)
+			if resources != "" {
+				return false, fmt.Errorf("Resources left running after stop:\n%s", resources)
```
Comment (mengqiy) [inline test/e2e/framework/util.go]: You should not return an error here, or the backoff func will stop earlier. See: https://github.com/kubernetes/apimachinery/blob/210bcde8e5a7715fe776460ca1df1a3f2738725e/pkg/util/wait/wait.go#L203
But you may want to keep the error somewhere.
I'd expect something like:
```
    var e error
	verifyCleanupFunc := func() (bool, error) {
        e = nil
		for _, selector := range selectors {
			resources := RunKubectlOrDie("get", "rc,svc", "-l", selector, "--no-headers", nsArg)
			if resources != "" {
                 e = fmt.Errorf("Resources left running after stop:\n%s", resources)
				return false, nil
			}
			pods := RunKubectlOrDie("get", "pods", "-l", selector, nsArg, "-o", "go-template={{ range .items }}{{ if not .metadata.deletionTimestamp }}{{ .metadata.name }}{{ \"\\n\" }}{{ end }}{{ end }}")
			if pods != "" {
                 e = fmt.Errorf("Pods left unterminated after stop:\n%s", pods)
				return false, nil
			}
		}
		return true, nil
	}
	err := wait.ExponentialBackoff(backoff, verifyCleanupFunc)
	if err != nil {
         // if timeout, you want to return e or err + e.
         // err itself may not be very helpful.
		Failf(e.Error())
	}
```

Comment (Liujingfang1) [reply test/e2e/framework/util.go]: That's a good catch and thank you. I updated it.

File: test/e2e/framework/util.go [hunk 3]
```diff
@@ -2066,15 +2066,32 @@ func AssertCleanup(ns string, selectors ...string) {
...
+		Factor:   2,
+		Steps:    3,
+	}
+	var e error
+	verifyCleanupFunc := func() (bool, error) {
+		e = nil
+		for _, selector := range selectors {
+			resources := RunKubectlOrDie("get", "rc,svc", "-l", selector, "--no-headers", nsArg)
+			if resources != "" {
+				e = fmt.Errorf("Resources left running after stop:\n%s", resources)
```
Comment (soltysh) [inline test/e2e/framework/util.go]: In both cases here and few lines below you can just return the error with the false. This way the error will bubble up to `wait.ExponentialBackoff` caller. It'll be either your error or the timeout error returned by the wait, but in both cases you'll have a clear information what happened, better than the current.

**Review by soltysh — CHANGES_REQUESTED** [review]

File: test/e2e/framework/util.go (hunk 3 above)
Comment (mengqiy) [reply test/e2e/framework/util.go]: > you can just return the error with the false.

If you take a closer look at `wait.ExponentialBackoff`, you will see it won't retry at all when the condition function returns an error.

Comment (Liujingfang1) [reply test/e2e/framework/util.go]: @soltysh  In my previous change, the error was returned with false . Then Mengqi pointed out that `wait.ExponentialBackoff` will exit immediately when it sees an error, which causes the retry end early.

**Review by soltysh — APPROVED** [review]
/approve

**Pull request merged**